### ISSUE TITLE
Add Geocaching.SU support

### ISSUE THREAD

**Issue by okainov** [cgeo/cgeo#5269]
There is big ex-USSR geocaching community on http://www.geocaching.su/, it would be nice to have their database support.

**Comment by SammysHP** [comment]
Do they have an API?

See also arcao/Geocaching4Locus#51.

**Comment by rsudev** [comment]
That would also be my first question. In addition is interesting to know, whether their GPX is usable in c:geo or not, and what other features could be interesting and are accessible.

**Comment by GodfatherThe** [comment]
Yes, I've asked administrators and they provided me API and description, but it's in Russian. If there is anyone, who will code this functionality, I can work with this person as user of geocaching.su and as native Russian speaker. API is 9-page text document and I do not want to waste time to translate it without any actions further.
In short, API allows to:
Open functions:
- Get caches inside the given square
- Get caches around the given point
Also additional filtering by cache types can be applied.
- Get cache info
- Get photos
- Get caches by country
Authorization-only functions:
- Get personal caches (i.e. found by current person)
- Add note to e-notepad

**Comment by GodfatherThe** [comment]
@rsudev, about GPX - can you please give more details, what is the question? GPX is universal format, it shouldn't depend on application or anything else...

**Comment by rsudev** [comment]
@GodfatherThe GPX in its core does not support anything geocaching - and comes also in different versions. For geocaching, a very commonly used extension is the one from Groundspeak (geocaching.com), others frequently in use are e.g. from GSAK and there are some site specific ones.
It is also relatively common to have geocodes (waypoint names in GPX-speak) which start with a site-specific two-letter combination (GC for geocaching.com, OC for opencaching.de, ...) which would allow c:geo to easily detect which site a specific cache is for. So these are (some of) the questions around their GPX format.
Regarding the Api it would be interesting whether it also allows to post logs for a cache, how the authentication is done and what the data format it uses is, as this influences the amount of work associated with such a connector implementation.

**Comment by SammysHP** [comment]
I'm not a native Russian speaker and only understand some words, but a link to the API documentation would be helpful.

**Comment by GodfatherThe** [comment]
@rsudev I will try to investigate about GPX. Am I correct that this question is equal to question about output format for API functions? So function returning all caches in some square should return some GPX-like file?

As I wrote above, API allows to post logs ("add note" in my words). Authorization is based on OAuth 1.0.

@SammysHP 
I attached the document with API.
[APIV0.04.docx](https://github.com/cgeo/cgeo/files/36387/APIV0.04.docx)

**Comment by rsudev** [comment]
I produced a google-translated version of the document - relatively terse but understandable. Open questions are e.g. how you sign up for a key and an account (needed for development testing).


[api-doc-en-google-translate.docx](https://github.com/cgeo/cgeo/files/37294/api-doc-en-google-translate.docx)

**Comment by SammysHP** [comment]
Oh yes, that is important. The API must be public, that means everyone must be able to get a key for free.

**Comment by GodfatherThe** [comment]
When I asked administration about "start keys" they said that for start you can use consumer_key key from the document to get initial tokens and then use them for authorization.

**Comment by rsudev** [comment]
@Bananeweizen You talk abotu geocaching.ru in commit message and connector name - shouldn't it be geocaching.su not only for the host?

**Comment by Bananeweizen** [comment]
the domain works with .su and .ru. The about page explictly shows mails with .ru. However, I'm not sure which they prefer themselfes. To me RU (Russia) fits better than SU (Soviet Union), and the ICANN also wants to remove the SU domain.

**Comment by GodfatherThe** [comment]
BTW The main website is geocaching.su, so it's better to use this name

**Comment by GodfatherThe** [comment]
Geocaching.ru does just redirect to .su

**Comment by rsudev** [comment]
BTW @GodfatherThe What is your view on a Russian version of c:geo? There was a request for that, but right now the translation more or less stagnates at 18%. Do you think people would appreciate that, or do they rather use an English version?

**Comment by GodfatherThe** [comment]
I personally use English version, and for now I don't think there are a lot of Russian people using c:geo mainly because geocaching.com is not popular in ex-ussr at all and c:geo is mainly a client for .com. 
But if .su support will be added, then it would be really nice to have russian translation in addition to our caches support.

**Comment by Lineflyer** [comment]
> But if .su support will be added, then it would be really nice to have russian translation in addition to our caches support.

@GodfatherThe 
You can contribute to the russian translation on https://crowdin.com/project/cgeo
Your support would be highly appreciated.

**Comment by GodfatherThe** [comment]
Any updates? Is it about lack of time to develop or some issues with API?

**Comment by rsudev** [comment]
From side currently lack of time (and I assume the same for @Bananeweizen who worked on the GPX already)

**Comment by GodfatherThe** [comment]
@Bananeweizen is #5269 included to release version?
BTW if you need any API help or fixes from server side, please tell me, I'll try to implement it on server side.

**Comment by GodfatherThe** [comment]
Just installed nightly version - looks great! Thanks a lot for making it live!!! Hope, it will be in production quality in the next few months! Thank you!

Some issues I observe:
- Missing the most important part about how to find the cache, only description is downloaded. Talking about API mapping, seems you only use "adesc" field, but it's also needed to use "cdesc". And it would be great, if those two parts will be separated somehow in interface.
- Incorrect GeoCodes, it's not "SUXXXXX", it should be "TRXXXX"/"VIXXXX"/"MSXXXX"/"MVXXXX" depending on cache types (traditional\virtual etc)
- Incorrect notebook logs - all notes have type "Write note", but they're different in real.
- For some reason, on live map all caches look as a gray box with "??", correct icon appears only after saving cache for offline use.

**Comment by Lineflyer** [comment]
IMHO you should create dedicated issues for each bug you found (e.g. cdesc instead of adesc) and for each improvement proposal. This will help us solving it step by step as soon as a dev has time for it.

**Comment by GodfatherThe** [comment]
@Lineflyer I just don't know development status for it, maybe @Bananeweizen knows about some of them already (i.e. for notes types). If needed, I can post separated issues, no problems.

**Comment by Bananeweizen** [comment]
The additional status field for logs is not available in my responses, doesn't matter how many more characters I add to the "istr" parameter.

For the geocodes and the unknown type on the live map I'm preparing fixes. The unknown type was simply because the request didn't have an istr parameter to resolve types.

Regarding images I don't want to probe your server for images, therefore please add the cache hint image to the list of images, or create a new tag for it. Right now cgeo only knows hint images and log images (and shows only these on the images page of the cache details activity). That's why I added all of them to the hints. We could instead add them to the description (since that is essentially HTML text, and adding some img tags would be simple).

Generally I would like it if you could align your different XML outputs more. Depending on what I request, I get different root nodes and different nesting in the xml. That is not so nice for actually parsing xml (and we are parsing XML, not text). E.g. when requesting a single cache, its ID will be under data->id, while for multiple caches it will be under data->cache->id. I suggest that you create an XML schema, which fits for both, and really produce your output according to that schema in all cases. Similarly there are minor differences in how tags are named. The cache type is in type or ctype. The name of the cache owner is in nick or autor. Please unify all those.

You produce invalid XML files in some cases. Take http://www.geocaching.su/site/api.php?rtype=2&cid=1235&istr=abcdefghiklmnopsqrtuvwxyz as example. That should show an error in your browser. And it breaks our XML parser, leading to runtime exceptions.

**Comment by GodfatherThe** [comment]
About status - forgot to deploy it to production :), now I've checked it - should appear in all queries asking for notes.

About geocode - there is "m" option returning type in string form. Or do you need parameter returning exact geocode like "TR"\"MS"?

Images - as a quick fix, just added to images tag as a first element, but maybe will improve somehow later.

About unification - in fact, now we have another programmer who is working on brand new API. It's not in production yet, but maybe will be there in few months. 
I don't mind about changing current API as well, but I don't know who else uses it.... Because if there are some people, then structure changes and renamings will break all their applications.. I will ask on our forum about it..

There was issues with HTML tags inside some fields, fixed it for vpart, please report any other obervations if any. 
BTW in adesc,cdesc,vpart,tpart I've added escaped HTML tags (before, they were wiped out). Hope, you can display it in c:geo.

Thanks for the report with company - fixed, please try now.

**Comment by Bananeweizen** [comment]
what is status=0? http://www.geocaching.su/site/api.php?rtype=2&cid=9060&istr=abcdefghiklmnops, log of djin-sv.
I mapped 2 and 4 both to DIDNT_FIND, and 5 and 6 both to OWNER_MAINTENANCE. Either I didn't get the meaning from your description, or we can't map that any better to the available log types in cgeo.

cgeo will crash on archived caches http://www.geocaching.su/site/api.php?rtype=2&cid=9876&istr=abcdefghiklmnops as well as caches in moderation http://www.geocaching.su/site/api.php?rtype=2&cid=18858&istr=abcdefghiklmnops, since the API delivers lots of empty tags then. There should be some explicit information about this. Even the status field is empty, where I would have expected this.

All other issues should be fixed with the next nightly. If you have further ideas about improvements, you can simply create a pull request to be merged to our code. Might be easier then first describing the necessary changes.

**Comment by GodfatherThe** [comment]
status=0 looks like some ancient out-of-date note, I don't think there should be any of them, so for this case it looks like something exceptional. So "unknown" is ok.

Mapping looks good for me at the moment. The only thing I would like to check - №5, "repaired". For "reading mode" it's ok to be as "owner_maintenance". But what about write? I mean when I'm as a player visited the cache, found it robbed\broken\no container etc, and then I repaired it (restored container with my own, it's a recommended way in SU geocaching - to repair caches which are not active \ robbed). So I would like to have an ability to write note from c:geo with type "repaired". If other geocaching serviced don't have such type, it looks like Feature request I should submit separately, correct?

Archive and caches awaiting moderations should not be visible, I assume. Request asking for caches in area should not return any of them. It looks like another exceptional situation. But your concern is understandable, because we already return some piece of info about non-active caches. So I've added returning other information for non-active caches as well. But please do not use\ignore all caches with status != 1. I can assume the only user story when user saved cache and it was moved to archive several days\months later. Then user clicks "Refresh" and... I don't know what to do in such situation.. What do you do now when offline-saved cache was suddenly removed\archived on server side?

**Comment by GodfatherThe** [comment]
Just FYI we have new more neat API with normal doc http://www.geocaching.su/api/documentation.html
iOS app migrated to it and it looks good.

**Comment by Lineflyer** [comment]
I received a mail on support and forwarded to cgeo-dev mailing list.

**Comment by GodfatherThe** [comment]
The only note is that for the moment it works (should work) only with authorized users. If it's critical we can dicsuss it internally to make this limitation only for specific requests (i.e. send note), not for all

**Comment by Lineflyer** [comment]
There was recent activity on the IRC developer chat.
If that person reads this, please just follow up here...might be easier to get answers.

**Comment by GodfatherThe** [comment]
Okay, just thought IRC is a better place for it.
On behalf of geocaching.su community we'd really like to have this implemented (authorization support for SU + logging from the app). We're ready to support you financially exactly for this feature. Is it possible somehow, can we discuss it? Thanks.

**Comment by Lineflyer** [comment]
Financial support does not help at all, as we are no commercial team. We just need someone developing and contributing this feature into our opensource project under the known license.
Not sure how others see that, but if you want to hire someone doing this, feel free. However the team always judges on code quality and if the functionality behind fits into the opensource spirit.

**Comment by okainov** [comment]
Just a bit of granularity what is still missing and what we'd like to have here (I can create separated issues if needed)

- [ ]  Authorization. 
It's the most desired feature for us at the moment. I played with the code a bit, but seems there are no examples I can try to modify for our needs. Current API is OAuth 1.0 as described above, 3-legged token request + login. I was able to get successful login reply, but not with keys exchange.
- [ ] Hiding current user's found caches on the map
- [ ] Postings logs
- [ ] Postings photos
- [ ] Voting for cache
- [ ] Sending answers for Virtual caches. The is probably something unique to gc.su because I haven't seen it in other services. To be simple, it's just about sending some text as "answer to question in virtual cache". I'm not really sure how does it fit the general c:geo appearence
- [x] Waypoints support. I Implemented it in #7128 using the current parsing as baseline.

I would say that I might be able to implement the listed functions if there would be an example with some API calls I can use as a baseline. Currently I think I'm stuck on "authorization" and I'm not sure I can implement it by myself.
